torch/_inductor/codegen/cpp_gemm_template.py

```diff
@@ -0,0 +1,372 @@
+                    self.num_threads, factor, m_blocks, n_blocks, k_blocks
+                )
+            cofactor = self.num_threads // factor
+            if m_blocks % cofactor == 0:
+                return get_blocking(
+                    self.num_threads, factor, m_blocks, n_blocks, k_blocks
+                )
+        raise AssertionError("Should not reach here.")
+
+    @cache_on_self
+    def cache_blocking(self) -> GemmBlocking:
+        # TODO(jgong5): improve cache blocking with CPU info
+        assert (
+            not self.is_dynamic_M
+        ), "Unable to determine cache blocking for dynamic M."
+        thread_blocking = self.thread_blocking()
+        return GemmBlocking(thread_blocking.block_m, 1, thread_blocking.block_k)
+
+    @staticmethod
+    def add_choices(
+        choices, layout, input_nodes, beta=1, alpha=1, trans_w=False, input_indices=None
+    ):
+        if input_indices is None:
+            input_indices = list(range(len(input_nodes)))
+
+        def reorder_and_filter(inputs, layout_or_out):
+            if len(input_indices) == 2:
+                x_idx = input_indices[0]
+                w_idx = input_indices[1]
+                return [inputs[x_idx], inputs[w_idx]], layout_or_out
+            else:
+                assert (
+                    len(input_indices) == 3
+                ), "Cpp Packed GEMM template requires 2 or 3 input nodes."
+                # assume the input order is [inp, x, w] and we reorder it to [x, w, inp]
+                inp_idx = input_indices[0]
+                x_idx = input_indices[1]
+                w_idx = input_indices[2]
+                return [inputs[x_idx], inputs[w_idx], inputs[inp_idx]], layout_or_out
+
+        def transpose_weight(inputs, layout_or_out):
+            if not trans_w:
+                return inputs, layout_or_out
+
+            new_inputs = list(inputs)
+            W = inputs[1]
+            if isinstance(W, ir.IRNode):
+                if not isinstance(W, ir.TensorBox):
+                    W = ir.TensorBox(W)
+                new_inputs[1] = L.permute(W, [1, 0])
+                return new_inputs, layout_or_out
+            else:
+                assert isinstance(W, torch.Tensor)
+                new_inputs[1] = W.transpose(0, 1)
+            return new_inputs, layout_or_out
+
+        # TODO(jgong5): decide proper number of threads per problem size
+        num_threads = parallel_num_threads()
+        new_inputs, _ = transpose_weight(*reorder_and_filter(input_nodes, layout))
+        m, n, k, *_ = mm_args(new_inputs[0], new_inputs[1])
+        micro_gemm = create_micro_gemm(
+            "micro_gemm", m, n, k, layout.dtype, alpha=alpha, num_threads=num_threads
+        )
+        assert micro_gemm is not None
+        _, block_n, _ = micro_gemm.register_blocking
+
+        def pack_weight(inputs, layout_or_out):
+            W = inputs[1]
+            new_inputs = list(inputs)
+            if isinstance(W, ir.IRNode):
```

Comment:
May I ask a question, in which case will `isinstance(W, ir.IRNode)` is False? After we change W into a constant in `postprocessor`?

Reply:
`pack_weight` will be called in the autotune benchmark when `W` would be a real tensor hence `isinstance(W, ir.IRNode)` is `False`.